src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs

```diff
@@ -15,27 +16,33 @@ public class CompilationWithAnalyzers
         private readonly Compilation _compilation;
         private readonly CancellationToken _cancellationToken;
         private readonly ConcurrentSet<Diagnostic> _exceptionDiagnostics;
+        private ImmutableHashSet<SyntaxTree> _analyzedSyntaxTrees = ImmutableHashSet<SyntaxTree>.Empty;
+        private Dictionary<SyntaxTree, Task<ImmutableArray<Diagnostic>>> _documentTasks = new Dictionary<SyntaxTree, Task<ImmutableArray<Diagnostic>>>();
+        private Task<ImmutableArray<Diagnostic>> _completeAnalysisTask;
+        private Task _latestAnalysisTask;
+        private readonly object _analysisLock = new object();
 
         public Compilation Compilation
         {
             get { return _compilation; }
         }
-
+        
         /// <summary>
         /// Creates a new compilation by attaching diagnostic analyzers to an existing compilation.
         /// </summary>
         /// <param name="compilation">The original compilation.</param>
         /// <param name="analyzers">The set of analyzers to include in future analyses.</param>
         /// <param name="options">Options that are passed to analyzers.</param>
+        /// <param name="onAnalyzerException">Action to invoke if an analyzer throws an exception.</param>
         /// <param name="cancellationToken">A cancellation token that can be used to abort analysis.</param>
-        public CompilationWithAnalyzers(Compilation compilation, ImmutableArray<DiagnosticAnalyzer> analyzers, AnalyzerOptions options, CancellationToken cancellationToken)
+        public CompilationWithAnalyzers(Compilation compilation, ImmutableArray<DiagnosticAnalyzer> analyzers, AnalyzerOptions options, Action<Exception, DiagnosticAnalyzer, Diagnostic> onAnalyzerException, CancellationToken cancellationToken)
```

Comment:
Instead of adding a parameter to the existing constructor, a new overloaded constructor must be introduced. The compatibility council has ruled out changing the public API for this.